Zend/zend_object_handlers.c

```diff
@@ -710,6 +718,13 @@ ZEND_API zval *zend_std_read_property(zend_object *zobj, zend_string *name, int
 			}
 
 			zval_ptr_dtor(&tmp_result);
+
+			retval = zend_std_read_dynamic_property(zobj, name, cache_slot, property_offset);
```

Comment:
This code is also called for typed properties which were unset():

```
<?php
class A {
    public int $foo;

    public function __isset($prop) {
        echo __FUNCTION__, "\n";
        $this->$prop = 123;
        return true;
    }
    public function __get($prop) {
        throw new Exception('Unreachable');
    }
}

$a = new A;
unset($a->foo);
echo $a->foo ?? 234;
```

You cannot assume this is a dynamic property at that location.

Reply:
Looks like this is unsound too. For declared properties, we're returning a direct property into the object. If we call `__isset` the user might releat the object. Once we delref in before exiting `zend_std_read_property` the pointer will become invalid. This means we'll have to use `rv` for this case, but I don't know what impact that may have.

Reply:
Why would that pointer become invalid? It's still within the object. As long as the object itself isn't freed - which it won't within zend_std_read_property a declared property pointer will stay valid.

Reply:
@bwoebi The object can be freed by `__isset`. The problem is not within `zend_std_read_property`, but once we pass the pointer back to the VM. Should become evident in the patch. Just a sec.

Reply:
the zval pointer itself won't escape the lifetime of anyone holding a reference to the object ... I would expect?
 Anyway, I'll wait for your patch :-)

Reply:
Pushed the patch. You can try removing the `ZVAL_COPY(rv, retval); retval = rv;` to trigger the error I was referencing.

Reply:
Ah, I get what you mean. Yes, that's ugly :-D

Reply:
Indeed :smile: Let me know if you have any improvements in mind. Anyway, this has reaffirmed the decision to delay this for 8.4.